src/python/pants/jvm/resolve/coursier_fetch_integration_test.py

```diff
@@ -668,3 +668,60 @@ def test_transitive_excludes(rule_runner: RuleRunner) -> None:
     entries = resolve.entries
     assert any(i for i in entries if i.coord.artifact == "jackson-databind")
     assert not any(i for i in entries if i.coord.artifact == "jackson-core")
+
+
+@pytest.mark.xfail(
+    reason="coursier bug: https://github.com/coursier/coursier/issues/2884", strict=True
+)
+@maybe_skip_jdk_test
+def test_missing_entry_for_transitive_dependency(rule_runner: RuleRunner) -> None:
+    resolve = rule_runner.request(
+        CoursierResolvedLockfile,
+        [
+            ArtifactRequirements(
+                [
+                    Coordinate(
+                        group="org.apache.hive",
+                        artifact="hive-exec",
+                        version="1.1.0",
+                    )
+                    .as_requirement()
+                    .with_extra_excludes(
+                        "org.apache.calcite:calcite-avatica",
+                        "org.apache.calcite:calcite-core",
+                        "jdk.tools:jdk.tools",
+                    )
+                ]
+            )
+        ],
+    )
+
+    lookup = {(entry.coord.group, entry.coord.artifact): entry for entry in resolve.entries}
+    missing = []
+    for entry in resolve.entries:
+        for d in entry.dependencies:
+            coord = (d.group, d.artifact)
+            if coord not in lookup:
+                missing.append(coord)
+
+    # We expect all the dependencies to have an entry, but right now it's not true
+    # for ("junit", "junit") and ("org.apache.curator", "apache-curator").
+    # TODO Remove the workaround once the bug is fixed.
+    assert missing == []
+
+
+@pytest.mark.xfail(reason="coursier bug?", strict=True)
+@maybe_skip_jdk_test
+def test_failed_to_fetch_jar_given_packaging_pom(rule_runner: RuleRunner) -> None:
+    reqs = ArtifactRequirements(
+        [
+            Coordinate(
+                group="org.apache.curator",
+                artifact="apache-curator",
+                version="5.5.0",
+            ).as_requirement()
+        ]
+    )
+
+    # Exception: No jar found for org.apache.curator:apache-curator:5.5.0.
```

Comment:
Does this mean that an explicit `jvm_artifact(...)` referring to this `apache-curator` stops `pants generate-lockfiles` from working too?

Reply:
No, `pants generate-lockfiles` is working but then other goals stop working:
```
Engine traceback:
  in `test` goal

KeyError: ('org.apache.curator', 'apache-curator')
```
```
Engine traceback:
  in `package` goal

KeyError: ('org.apache.curator', 'apache-curator')
```

Reply:
Ah. That seems like it's worth a dedicated separate issue: pants should never be crashing with a raw exception like `KeyError`. Could you file one? Thanks.

Reply:
You were right, it fails on `generate-lockfiles`, here is an issue https://github.com/pantsbuild/pants/issues/20225